src/Compilers/CSharp/Portable/Symbols/MethodSymbolExtensions.cs

```diff
@@ -190,6 +191,48 @@ public static bool IsAsyncReturningIAsyncEnumerator(this MethodSymbol method, CS
                 && method.ReturnType.IsIAsyncEnumeratorType(compilation);
         }
 
+#nullable enable
+        public static bool IsAsyncReturningTaskViaOverride(this MethodSymbol method, [NotNullWhen(true)] out object? builderArgument)
+        {
+            builderArgument = null;
+            return method.IsAsync
+                && method.HasMethodLevelBuilder(builderArgument: out builderArgument)
+                && method.ReturnType is NamedTypeSymbol { Arity: 0 };
+        }
+
+        public static bool IsAsyncReturningGenericTaskViaOverride(this MethodSymbol method, [NotNullWhen(true)] out object? builderArgument)
+        {
+            builderArgument = null;
+            return method.IsAsync
+                && method.HasMethodLevelBuilder(builderArgument: out builderArgument)
+                && method.ReturnType is NamedTypeSymbol { Arity: 1 };
+        }
+
+        /// <summary>
+        /// Returns true if the method has a [AsyncMethodBuilder(typeof(B))] attribute. If so it returns the "B".
+        /// Validation of builder type B is left for elsewhere. This method returns B without validation of any kind.
+        /// </summary>
+        internal static bool HasMethodLevelBuilder(this MethodSymbol method, [NotNullWhen(true)] out object? builderArgument)
```

Comment:
>HasMethodLevelBuilder(this MethodSymbol method [](http://example.com/codeflow?start=29&length=46)

Consider changing this to an extension method on `Symbol` and call this method from `TypeSymbolExtensions.IsCustomTaskType()` as well.

Reply:
I'm looking at one usage of `IsCustomTaskType` in `BoundLambda.CalculateReturnType` which I need to think more about.

Reply:
I didn't understand the comment about `CalculateReturnType`. Could we share this implementation with `TypeSymbolExtensions.IsCustomTaskType()`?

Reply:
The comment about `CalculateReturnType` is no longer relevant now that we're checking the builder from the return type.
Factored logic with `IsCustomTaskType`.